### PR TITLE
refactor(x/data): update data module to use orm (part 2)

### PR DESCRIPTION
**Pull request by ryanchristo**
## Description

Closes: #248
Closes: #897

This pull request completes the migration from KVStore to orm tables in the data module.

In this pull request I propose `hasher` in place of the lookup table, which takes the same configuration options but is specifically used for generating shortened identifiers.

Updating the lookup table to use orm tables would have required a more robust configuration or restricted its use to the DataID orm table. The number of collisions are tracked outside `hasher` and passed as a parameter to `CreateID`. I'm not sure if this is the best approach but it reduces the complexity of integrating orm tables.

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/data/server/server.go [hunk 1]
```diff
@@ -24,12 +29,12 @@ var ModuleSchema = ormv1alpha1.ModuleSchemaDescriptor{
...
 type serverImpl struct {
 	storeKey   sdk.StoreKey
-	iriIDTable lookup.Table
+	iriHasher  hasher.Hasher
 	stateStore api.StateStore
 }
 
 func newServer(storeKey sdk.StoreKey) serverImpl {
-	tbl, err := lookup.NewTable([]byte{IriIDTablePrefix})
+	hasher, err := hasher.NewHasher([]byte{IriHashPrefix})
```
Comment (ryanchristo) [inline x/data/server/server.go:37]: As mentioned above, do we still need a prefix if we are using orm tables?

File: x/data/server/utils.go [hunk 2]
```diff
@@ -1,50 +0,0 @@
...
-	if err != nil {
-		return nil, err
-	}
-
-	timestamp := &gogotypes.Timestamp{
-		Seconds: dataAnchor.Timestamp.Seconds,
-		Nanos:   dataAnchor.Timestamp.Nanos,
-	}
-
-	iri := string(s.iriIDTable.GetValue(store, id))
```
Comment (ryanchristo) [inline x/data/server/utils.go:37]: This was the only use of `GetValue` using the lookup table but we store the id in each orm table and we can get the IRI from an id and the id from an IRI (i.e. the orm table already supports reverse lookups).

This makes me question whether we even need to hash the data id... Why not use an auto-incrementing id as we do for other orm tables? Is it still necessary to generate our own compact id?

When anchoring data, we anchor using the content hash. In the current implementation, we then generate a compact id for the anchored data and only allow queries by IRI and content hash. Verifying the compact id with a reverse lookup only verifies that we properly generated the id and stored it in the orm table.

Is there another use case? When would we need to verify the compact id?

**Review by ryanchristo — COMMENTED** [review]
I have several questions, the first and foremost being do we need compact ids? Why not use an auto-incrementing id as we do for other orm tables? A user never provides the id, only an IRI or a content hash. If we verify the id, we are only really verifying that we properly generated and stored the id. Is there another use case for the compact id and to support reverse lookups?

File: x/data/server/hasher/doc.go [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+using an efficient, non-cryptographic hash function.
+
+A new Hasher instance can be created with the NewHasher() function. Advanced users can use
+the NewHasherWithOptions() function to tweak the underlying parameters, but the defaults
+were chosen based on testing and should provide a good balance of performance and storage
+efficiency.
+
+Shortened identifiers are generated using the idempotent Hasher.CreateID method.
+
+Using the default algorithm which uses the first 4 bytes of a 64-bit FNV-1a hash and then
```
Comment (aaronc) [inline x/data/server/hasher/doc.go:12]: Let's just use a 64bit blake2b or blake2s

File: x/data/server/hasher/hasher.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+// HashOptions is used to specify custom hash options and should only be used by advanced users.
+type HashOptions struct {
+	// NewHash is a function which returns a new hash.Hash instance.
+	NewHash func() hash.Hash
+
+	// MinLength is the minimum number of hash bytes that will be used to create a lookup identifier.
+	MinLength int
+
+	// Prefix is an optional prefix to be pre-pended to all keys.
+	Prefix []byte
```
[the remark this answers is not in the document]
Comment (aaronc) [reply x/data/server/hasher/hasher.go:72]: 🤷

File: x/data/server/msg_anchor.go [hunk 5]
```diff
@@ -37,14 +37,33 @@ func (s serverImpl) anchorAndGetIRI(ctx context.Context, ch ToIRI) (iri string,
 		return "", nil, nil, err
 	}
 
-	sdkCtx := sdk.UnwrapSDKContext(ctx)
-	store := sdkCtx.KVStore(s.storeKey)
-	id = s.iriIDTable.GetOrCreateID(store, []byte(iri))
+	dataId := &api.DataID{Iri: ""}
+
+	for collisions := 0; dataId.Iri != iri; collisions++ {
```
Comment (aaronc) [inline x/data/server/msg_anchor.go:42]: would be good to at least break this out into a standalone function

File: x/data/server/server.go (hunk 1 above)
Comment (aaronc) [reply x/data/server/server.go:37]: no

**Review by aaronc — COMMENTED** [review]
Looking like the right direction 👍 . Pre-approving. I'd like the next reviewer to do a more thorough look at the tests

**Review by aaronc — APPROVED** [review]

File: x/data/server/query_data.go [hunk 6]
```diff
@@ -6,22 +6,20 @@ import (
 	"github.com/cosmos/cosmos-sdk/orm/model/ormlist"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
-
+	gogotypes "github.com/gogo/protobuf/types"
```
Comment (technicallyty) [inline x/data/server/query_data.go:9]: ඞ this import looks sus ඞ

File: x/data/server/query_data.go [hunk 7]
```diff
@@ -107,3 +113,24 @@ func (s serverImpl) ByAttestor(ctx context.Context, request *data.QueryByAttesto
...
+
+	dataAnchor, err := s.stateStore.DataAnchorTable().Get(ctx, dataId.Id)
+	if err != nil {
+		return nil, err
+	}
+
+	return &data.ContentEntry{
+		Hash: ch,
+		Iri:  iri,
+		Timestamp: &gogotypes.Timestamp{
```
Comment (technicallyty) [inline x/data/server/query_data.go:131]: this pattern of protobuf ts -> gogo ts seems to happen in quite a few places now, maybe in another PR or this one we can make a util for it?

**Review by technicallyty — APPROVED** [review]
lgtm, 2 nits

File: x/data/server/hasher/doc.go (hunk 3 above)
Comment (ryanchristo) [reply x/data/server/hasher/doc.go]: https://github.com/regen-network/regen-ledger/pull/970/commits/c95c5158524f5da2105565bd8472e70686293dfc

File: x/data/server/query_data.go (hunk 6 above)
Comment (ryanchristo) [reply x/data/server/query_data.go]: https://github.com/regen-network/regen-ledger/pull/970/commits/2f202a60b2799a65b05cf8d09b9d0c3657155062

File: x/data/server/msg_anchor.go (hunk 5 above)
Comment (ryanchristo) [reply x/data/server/msg_anchor.go:56]: https://github.com/regen-network/regen-ledger/pull/970/commits/16b101c0f8b09fb5f8db7232707d2c8998d07781

File: x/data/server/server.go (hunk 1 above)
Comment (ryanchristo) [reply x/data/server/server.go]: https://github.com/regen-network/regen-ledger/pull/970/commits/36f63e20da4bb0df865d43f342ebced627650bc5

File: x/data/server/hasher/hasher.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ryanchristo) [reply x/data/server/hasher/hasher.go]: https://github.com/regen-network/regen-ledger/pull/970/commits/36f63e20da4bb0df865d43f342ebced627650bc5

File: x/data/server/query_data.go (hunk 7 above)
Comment (ryanchristo) [reply x/data/server/query_data.go:133]: #981

**Comment by ryanchristo** [comment]
> I'd like the next reviewer to do a more thorough look at the tests

I'll start working on data acceptance tests.

**Comment by aaronc** [comment]
> > I'd like the next reviewer to do a more thorough look at the tests
> 
> I'll start working on data acceptance tests.

Ideally we do those at the same time that we write the proto files before implementation. Can we maybe start slowly shifting towards that approach, at least for v5?

**Comment by ryanchristo** [comment]
> Ideally we do those at the same time that we write the proto files before implementation. Can we maybe start slowly shifting towards that approach, at least for v5?

I completely agree. We should require acceptance tests before implementation starting with v5.0 updates.

File: x/data/server/hasher/hasher.go [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+func NewHasherWithOptions(options HashOptions) (Hasher, error) {
+	minLength := options.MinLength
+	if minLength == 0 {
+		minLength = 4
+	}
+
+	newHash := options.NewHash
+	if newHash == nil {
+		newHash = func() hash.Hash {
+			hash, err := blake2b.New512(nil)
```
Comment (aaronc) [inline x/data/server/hasher/hasher.go:38]: why 512? i think this should be
```suggestion
			hash, err := blake2b.New(8, nil)
```

**Review by aaronc — CHANGES_REQUESTED** [review]

File: x/data/server/hasher/hasher.go (hunk 8 above)
Comment (ryanchristo) [reply x/data/server/hasher/hasher.go]: Updated in https://github.com/regen-network/regen-ledger/pull/970/commits/425bb45ea5077b4ec0664c9d59cec0c62386a3c7

**Review by aaronc — APPROVED** [review]

**Pull request merged**